File: app/gilded-rose.ts
export class Item {
    name: string;
    sellIn: number;
    quality: number;

    constructor(name, sellIn, quality) {
        this.name = name;
        this.sellIn = sellIn;
        this.quality = quality;
    }
}

export class GildedRose {
    items: Array<Item>;

    constructor(items = [] as Array<Item>) {
        this.items = items;
    }

    updateQuality() {
        for (let i = 0; i < this.items.length; i++) {
            const item = this.items[i];
            const isConjured = item.name.toLowerCase().includes("conjured");

            if (item.name !== 'Aged Brie' && item.name !== 'Backstage passes to a TAFKAL80ETC concert') {
                if (item.quality > 0) {
                    if (item.name !== 'Sulfuras, Hand of Ragnaros') {
                        const degradation = isConjured ? 2 : 1;
                        item.quality -= degradation;
                        if (item.quality < 0) item.quality = 0;
                    }
                }
            } else {
                if (item.quality < 50) {
                    item.quality++;
                    if (item.name === 'Backstage passes to a TAFKAL80ETC concert') {
                        if (item.sellIn < 11 && item.quality < 50) {
                            item.quality++;
                        }
                        if (item.sellIn < 6 && item.quality < 50) {
                            item.quality++;
                        }
                    }
                }
            }

            if (item.name !== 'Sulfuras, Hand of Ragnaros') {
                item.sellIn--;
            }

            if (item.sellIn < 0) {
                if (item.name !== 'Aged Brie') {
                    if (item.name !== 'Backstage passes to a TAFKAL80ETC concert') {
                        if (item.quality > 0) {
                            if (item.name !== 'Sulfuras, Hand of Ragnaros') {
                                const degradation = isConjured ? 2 : 1;
                                item.quality -= degradation;
                                if (item.quality < 0) item.quality = 0;
                            }
                        }
                    } else {
                        item.quality = 0;
                    }
                } else {
                    if (item.quality < 50) {
                        item.quality++;
                    }
                }
            }
        }

        return this.items;
    }
}

File: gilded_rose.py
# -*- coding: utf-8 -*-

class GildedRose(object):

    def __init__(self, items):
        self.items = items
        for it in self.items:
            self.clip(it)

    def clip(self, it):
        if it.name == "Sulfuras, Hand of Ragnaros":
            return

        if it.quality < 0:
            it.quality = 0
        elif it.quality > 50:
            it.quality = 50

    def handle_special_item(self, item) -> bool:
        item.sell_in -= 1

        # handle aged brie
        if item.name == "Aged Brie":
            item.quality += 1
            return True
        # handle passes
        elif "Backstage passes" in item.name:
            # apply rules for backstage passes quality
            if item.sell_in <= 0:
                item.quality = 0
            elif item.sell_in <= 5:
                item.quality += 3
            elif item.sell_in <= 10:
                item.quality += 2
            else:
                item.quality += 1

            return True
        # handle sulfuras
        elif item.name == "Sulfuras, Hand of Ragnaros":
            item.sell_in += 1
            return True
        else:
            return False

    def handle_regular_item(self, item) -> None:
        item.sell_in -= 1
        expired_multiplier = 1

        if item.sell_in  < 0:
            expired_multiplier = 2
        
        if "Conjured" in item.name:
            item.quality -= 2 * expired_multiplier
        else:
            item.quality -= 1 * expired_multiplier

    def update_quality(self) -> None:
        for it in self.items:
            is_special = self.handle_special_item(it)

            if is_special:
                pass
            else:
                self.handle_regular_item(it)

            self.clip(it)

class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)
